TESTS/mbed_drivers/timeout/timeout_tests.h

```diff
@@ -0,0 +1,416 @@
+ */
+template<typename T>
+void test_multiple(void)
+{
+    volatile uint32_t callback_count = 0;
+    T timeouts[NUM_TIMEOUTS];
+    for (size_t i = 0; i < NUM_TIMEOUTS; i++) {
+        timeouts[i].attach_callback(mbed::callback(cnt_callback, &callback_count), TEST_DELAY_US);
+    }
+    Thread::wait(TEST_DELAY_MS + 1);
+    TEST_ASSERT_EQUAL(NUM_TIMEOUTS, callback_count);
+}
+
+/** Template for tests: zero delay
+ *
+ * Test zero delay
+ * Given a Timeout object
+ * When a callback is attached with 0.0 s delay, with @a attach()
+ * Then the callback is called instantly
+ *
+ * Test zero delay
+ * Given a Timeout object
+ * When a callback is attached with 0.0 s delay, with @a attach_us()
+ * Then the callback is called instantly
+ */
+template<typename T>
+void test_no_wait(void)
+{
+    Semaphore sem(0, 1);
+    T timeout;
+    timeout.attach_callback(mbed::callback(sem_callback, &sem), 0ULL);
+
+    int32_t sem_slots = sem.wait(0);
+    TEST_ASSERT_EQUAL(1, sem_slots);
+    timeout.detach();
+}
+
+/** Template for tests: accuracy of timeout delay
+ *
+ * Test delay accuracy
+ * Given a Timeout object with a callback attached with @a attach()
+ * When the callback is called
+ * Then elapsed time matches given delay
+ *
+ * Test delay accuracy
+ * Given a Timeout object with a callback attached with @a attach_us()
+ * When the callback is called
+ * Then elapsed time matches given delay
+ */
+template<typename T, us_timestamp_t delay_us, us_timestamp_t delta_us>
+void test_delay_accuracy(void)
+{
+    Semaphore sem(0, 1);
+    T timeout;
+    Timer timer;
+
+    timer.start();
+    timeout.attach_callback(mbed::callback(sem_callback, &sem), delay_us);
+
+    int32_t sem_slots = sem.wait(osWaitForever);
+    timer.stop();
+    TEST_ASSERT_EQUAL(1, sem_slots);
+    TEST_ASSERT_UINT64_WITHIN(delta_us, delay_us, timer.read_high_resolution_us());
+
+    timeout.detach();
+}
+
+#if DEVICE_SLEEP
+/** Template for tests: timeout during sleep
+ *
+ * Test timeout during sleep
+ * Given a Timeout object with a callback attached with @a attach()
+ *     and the uC in a sleep mode
+ * When given delay elapses
+ * Then the callback is called
+ *     and elapsed time matches given delay
+ *
+ * Test timeout during sleep
+ * Given a Timeout object with a callback attached with @a attach_us()
+ *     and the uC in a sleep mode
+ * When given delay elapses
+ * Then the callback is called
+ *     and elapsed time matches given delay
+ */
+template<typename T, us_timestamp_t delay_us, us_timestamp_t delta_us>
+void test_sleep(void)
+{
+    Semaphore sem(0, 1);
+    T timeout;
+    Timer timer;
+
+    timer.start();
+    sleep_manager_lock_deep_sleep();
+    timeout.attach_callback(mbed::callback(sem_callback, &sem), delay_us);
+
+    bool deep_sleep_allowed = sleep_manager_can_deep_sleep();
+    TEST_ASSERT_FALSE_MESSAGE(deep_sleep_allowed, "Deep sleep should be disallowed");
+    sleep();
+
+    int32_t sem_slots = sem.wait(osWaitForever);
```

Comment:
If this line is needed, then it should be moved outside timer start-stop section and wait value should be set to 0.